### PR TITLE
Auth wip

### PR DESCRIPTION
**Pull request by macias2137**

### CODE REVIEW

File: lib/vox/user_manager.ex [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      nil -> Argon2.no_user_verify()
+      {:error, :invalid_credentials}
+
+      user ->
+        if Argon2.verify_pass(password, user.password_hash) do
+          {:ok, user}
+        else
+          {:error, :invalid_credentials}
+        end
+    end
```
Comment (boksadawid) [inline lib/vox/user_manager.ex]: looks like a good use case for `with`

File: lib/vox/user_manager/error_handler.ex [hunk 2]
```diff
@@ -0,0 +1,14 @@
+defmodule Vox.UserManager.ErrorHandler do
+  import Plug.Conn
+  import Phoenix.Controller
+
+  @behaviour Guardian.Plug.ErrorHandler
+
+  @impl Guardian.Plug.ErrorHandler
+  def auth_error(conn, {type, _reason}, _opts) do
+    body = to_string(type)
```
Comment (boksadawid) [inline lib/vox/user_manager/error_handler.ex]: do you need this variable here?

File: lib/vox_web/controllers/restaurants_controller.ex [hunk 3]
```diff
@@ -1,28 +1,43 @@
...
   end
 
   def update(conn, %{"id" => id}) do
     restaurant = Restaurants.get_restaurant_by_id(id)
-    Restaurants.add_vote_to_restaurant(restaurant)
-    redirect(conn, to: Routes.restaurants_path(conn, :index))
+    voter_ids = Votes.get_voter_ids()
+    if conn.assigns.current_user.id not in voter_ids do
+      Votes.create_vote(%{user_id: conn.assigns.current_user.id, restaurant_id: restaurant.id})
+    end
```
Comment (boksadawid) [inline lib/vox_web/controllers/restaurants_controller.ex]: IMHO the Votes.create_vote should contain all logic with checking if user already voted and if not creating the vote

Comment (boksadawid) [reply lib/vox_web/controllers/restaurants_controller.ex]: also Votes.get_voter_ids() gets all votes, maybe we would need to query only votes with specific restaurant id and user id, and check if exists?

**Review by boksadawid — CHANGES_REQUESTED** [review]

File: lib/vox/user_manager/error_handler.ex (hunk 2 above)
Comment (macias2137) [reply lib/vox/user_manager/error_handler.ex]: it's an artefact, can be deleted

File: lib/vox_web/controllers/restaurants_controller.ex (hunk 3 above)
Comment (macias2137) [reply lib/vox_web/controllers/restaurants_controller.ex]: we need to check if a user has already voted for any one of the restaurants so checking only a certain user_id - restaurant_id combination would not give us the answer for that

Comment (macias2137) [reply lib/vox_web/controllers/restaurants_controller.ex]: as agreed, as all votes have a unique user_id anyway, I wrote a function user_voted?(id) in Votes context which uses Repo.exists? and checks if user already cast a vote. get_voter_ids is no longer necessary then

File: lib/vox/user_manager.ex (hunk 1 above)
Comment (macias2137) [reply lib/vox/user_manager.ex]: fixed

File: lib/vox_web/controllers/restaurants_controller.ex [hunk 4]
```diff
@@ -1,28 +1,42 @@
...
   end
 
   def update(conn, %{"id" => id}) do
     restaurant = Restaurants.get_restaurant_by_id(id)
-    Restaurants.add_vote_to_restaurant(restaurant)
-    redirect(conn, to: Routes.restaurants_path(conn, :index))
+    user_id = conn.assigns.current_user.id
+    unless Votes.user_voted?(user_id) do
+      Votes.create_vote(%{user_id: user_id, restaurant_id: restaurant.id})
+    end
```
Comment (boksadawid) [inline lib/vox_web/controllers/restaurants_controller.ex]: this whole logic should be inside the context

Comment (boksadawid) [reply lib/vox_web/controllers/restaurants_controller.ex]: so create_vote calls the user_voted? instead doing it inside the controller

**Review by boksadawid — CHANGES_REQUESTED** [review]

File: lib/vox_web/controllers/restaurants_controller.ex (hunk 4 above)
Comment (macias2137) [reply lib/vox_web/controllers/restaurants_controller.ex]: I realised I didn't fix this yesterday, working on it

Comment (macias2137) [reply lib/vox_web/controllers/restaurants_controller.ex]: ok this should do:

**Review by boksadawid — APPROVED** [review]

**Pull request merged**